src/msgAggregatorWorker.ts

```diff
@@ -84,7 +86,7 @@ export const parseSerialMessages = (
         });
       } else {
         // otherwise they are spaces
-        const values = message.split(/\s/);
+        const values = message.split(delimiterRegex);
```

Comment:
There is a regression here for a use case that is not explicitly supported by the specification, but is supported by the Arduino IDE 1.x Serial Plotter and previous implementation of this project. That use case is the combination of a trailing delimiter and `\r\n` separator.

For example:

```cpp
void setup() {
  Serial.begin(9600);
}
void loop() {
  for (byte value = 0; value < 4; value++) {
    for (byte variableOffset = 0; variableOffset < 3; variableOffset++) {
      Serial.print(value + variableOffset);
      Serial.print('\t');
    }
    Serial.println();
  }
  delay(100);
}
```

You can see how code that generates data via a `for` loop favors a trailing delimiter, and of course `Serial.println();` is a convenient way to produce a separator.

Previously, that sketch produced 3 variables as expected:

Arduino IDE 1.8.19:

![image](https://user-images.githubusercontent.com/8572152/188775082-ead87519-2cf0-420f-9d89-fbcfd72632d7.png)

Arduino IDE 2.0.0-rc9.3:

![image](https://user-images.githubusercontent.com/8572152/188775235-7caef3e9-83bd-413c-9d26-0d59238641b0.png)

After the change made here, it produces 4 variables:

![image](https://user-images.githubusercontent.com/8572152/188774020-0632851e-6e7a-4085-988f-96fcd3140a43.png)

The problem is the `\r` in the `\r\n` separator produced by `Serial.println();` is being treated as a separate value. Previously, that `\r` was consumed by `message.split(/\s/);` (because [it matches `/\s/`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Character_Classes#types)).

Reply:
What if the separator regex is modified to include the carriage return character `\r`? Would that solve the problem or am I still missing something ?
```
const separator = "\r?\n";
```

Is there a particular reason why `Serial.println` outputs a `\r\n` and not `\n` ? We cannot have two different conventions for EOL character. Either we amend the SerialPlotter protocol or drop support for `\r\n` entirely.

Reply:
```typescript
const separator = "\r?\n";
```

That general approach should work. However, `separator` is used in the code directly as a string in addition to its use in the definition of `separatorRegex`, so the implementation is not quite so simple.

It will need to be something like this (not tested):

```diff
diff --git a/src/msgAggregatorWorker.ts b/src/msgAggregatorWorker.ts
index b05dddf..913915b 100644
--- a/src/msgAggregatorWorker.ts
+++ b/src/msgAggregatorWorker.ts
@@ -20,7 +20,7 @@ let buffer = "";
 let discardFirstLine = true;
 const separator = "\n";
 const delimiter = "[, \t]+"; // Serial Plotter protocol supports Comma, Space & Tab characters as delimiters
-var separatorRegex = new RegExp(`(${separator})`, "g");
+var separatorRegex = new RegExp(/(\r?\n)/, "g");
 var delimiterRegex = new RegExp(`(${delimiter})`, "g");
 
 export const parseSerialMessages = (
@@ -55,7 +55,7 @@ export const parseSerialMessages = (
   // remove the previous buffer
   buffer = "";
   // check if the last message contains the delimiter, if not, it's an incomplete string that needs to be added to the buffer
-  if (messagesAndBuffer[messagesAndBuffer.length - 1] !== separator) {
+  if (!separatorRegex.test(messagesAndBuffer[messagesAndBuffer.length - 1])) {
     buffer = messagesAndBuffer[messagesAndBuffer.length - 1];
     messagesAndBuffer.splice(-1);
   }
@@ -65,7 +65,7 @@ export const parseSerialMessages = (
 
   // for each line, explode variables
   messagesAndBuffer
-    .filter((message) => message !== separator)
+    .filter((message) => !separatorRegex.test(message))
     .forEach((message) => {
       const parsedLine: { [key: string]: number } = {};
```

>We cannot have two different conventions for EOL character.

Why not? Inconsistency in EOL is a common situation. Not so long ago, Windows, Linux, and macOS each used a different EOL.

>Is there a particular reason why `Serial.println` outputs a `\r\n` and not `\n` ?

No idea. It has been that way since 2006, and in the proud tradition of Arduino the person who made the change didn't bother to explain the reason:

https://github.com/arduino/Arduino/commit/650c2752c892f63a21f9ee0a76eb395246bac475

>amend the SerialPlotter protocol

I think it is a good idea. The protocol already does specify that `\r\n` is supported by using `Serial.println()` in the all the example programs, but that is too vague. It should be stated explicitly that both are supported.

Reply:
>drop support for `\r\n` entirely.

It would be too disruptive. I'm certain the code can be made to support either without any real difficulty.

The only real challenge here is that there is no test coverage in this project, despite the fact that this code is very easy to write unit tests for.

Reply:
I would like to contribute by writing the tests but I still haven't managed to setup my dev environment right.

Ideal case, I would like to debug the modifications before committing it to the repo. However, I have not been able to interface my Arduino with the standalone web-app or write a websocket script to feed values to the web-app.

Currently, I've been modifying the randomValueGenerator code manually to feed various types of messages and check its results. If you could guide me in setting this up, I would be able to contribute a lot more.

Reply:
> ```ts
> const separator = "\r?\n";
> ```
> 
> That general approach should work. 

Do we have to support `\r` as an EOL character? In that case, the regex would be
```ts
const separator = "\r?\n?";
```

Reply:
>Do we have to support \r as an EOL character?

No. It has never been supported and I haven't seen any complaints about the lack of support (I think it is not a very common separator in Arduino sketches despite being supported in the line ending menu of Serial Plotter and Serial Monitor)

For example, this sketch:

```cpp
void setup() {
  Serial.begin(9600);
}
void loop() {
  for (byte value = 0; value < 4; value++) {
    for (byte variableOffset = 0; variableOffset < 3; variableOffset++) {
      Serial.print(value + variableOffset);
      Serial.print('\t');
    }
    Serial.print('\r');
  }
  delay(100);
}
```

Produces no output in the plotter:

arduino-serial-plotter-webapp@0.1.0

![image](https://user-images.githubusercontent.com/8572152/188826973-f7f8f698-f702-477b-a415-f84aa58c4b03.png)

Arduino IDE 1.8.19

![image](https://user-images.githubusercontent.com/8572152/188827382-6fad8131-c9c5-4047-8b03-7ecf452d9635.png)